shark-ai/export_and_serve.py

```diff
@@ -0,0 +1,412 @@
+            logger.error("Failed to compile sharded IR")
+            print(compile_subp.stderr.decode(), file=sys.stderr)
+            exit(compile_subp.returncode)
+
+        logger.info(
+            f"Successfully compiled and saved TP8 VMFB module to {tp8_vmfb_path}"
+        )
+
+
+def main(args: Namespace) -> None:
+    compile = args.compile
+    artifacts_dir = (
+        args.artifact_dir
+        if args.artifact_dir is not None
+        else Path(os.path.join(args.export_dir, "artifacts/"))
+    )
+    shard = args.shard
+    export = args.export
+    weight_loc = args.weight_file
+    export_dir = args.export_dir
+    ir_path = args.ir
+    tp = args.tensor_parallel
+
+    if ir_path is not None and not compile:
+        logger.warning(
+            "Path to IR was provided through -i, but compilation was not enabled."
+        )
+
+    batch_sizes = [1, 4]
+    if args.batch_sizes:
+        batch_sizes = [int(b.strip()) for b in args.batch_sizes.split(",")]
+
+    if args.verbose:
+        logger.setLevel(logging.INFO)
+
+    base_pipeline = BasePipeline(
+        compile,
+        shard,
+        export,
+        tp,
+        batch_sizes,
+        ir_path,
+        artifacts_dir,
+        weight_loc,
+        export_dir,
+    )
+
+    base_pipeline.exec_pipeline()
+
+
+def _get_argparser() -> CliParser:
+    msg = "Utility script to combine shark-ai tools"
+    # The argparser should not print the usage when an error occurs.
+    # We handle that ourselves.
+    parser = CliParser(
+        prog="export_and_serve.py",
+        description=msg,
+        formatter_class=RawTextHelpFormatter,
+    )
+    parser.add_argument(
+        "-v",
+        "--verbose",
+        action="store_true",
+        default=False,
+        help="set logging level to INFO. The default logging level is WARNING.",
+    )
+    parser.add_argument(
+        "-c",
+        "--compile",
+        action="store_true",
+        default=False,
+        help="compile the exported model as part of the pipeline. Default is FALSE.",
+    )
+    parser.add_argument(
+        "export_dir",
+        type=Path,
+        help="the directory where the exported artifacts will be saved.",
+    )
+    parser.add_argument(
+        "-w",
+        "--weight-file",
+        type=Path,
+        help="the location of the GGUF/IRPA file(s) that contain the parameters.",
+        required=True,
+    )
+    parser.add_argument(
+        "-e",
+        "--export",
+        action="store_true",
+        default=False,
+        help="export the model in tp1 mode.",
+    )
+    parser.add_argument(
+        "-a",
+        "--artifact-dir",
+        type=Path,
+        help="the location where the artifacts (sharded weights) should be saved. Defaults to EXPORT_DIR/artifacts/",
+    )
+    parser.add_argument(
+        "-s",
+        "--shard",
+        action="store_true",
+        help="shard the weight file in tp8 mode and export to MLIR.",
+    )
```

Comment:
Seems like `--export` and `--shard` could be replaced with a single `--shard_count` option? Would values other than 1 and 8 make sense? What about 2, 4, or 16?

If the options are kept, should they be mutually exclusive? The branches in the code seem to suggest that both could be set.